### PR TITLE
Pawn promotion

### PR DESCRIPTION
**Pull request by smlovin2**
- Adds several tests to test the controller, game model and some acceptance tests
- Starts to implement controller and game model code
- Adds a route for the promote_pawn action in the game controller
- Adds modal which is toggled from ajax move call
- Finishes up implementation of controller and model pawn promotion action and method
- Creates ajax call after submitting the promotion modal
- Still a work in progress (code is kind of a mess and can't get modal to close after submitting)

### CODE REVIEW

**Comment by smlovin2** [comment]
@mscoutermarsh Hey Mike, was really hoping to get some help with this one. The back end stuff I think should be ok but the jquery front end stuff seems kind of messy. I got something working at least but I feel like there has to be a much better way of doing this than what I have so if you have any suggestions on how to improve this code I would really appreciate it. Also, I can't get the modal to close after clicking the 'Promote'(aka submit) button. I've tried several things but nothing seems to be working so if I could get some help with that as well that would be great! Thanks!

File: app/views/games/show.html.erb [hunk 1]
```diff
@@ -7,14 +16,39 @@
...
+    $('#promote-form').submit(function(event) {
+      var piece_type = $('input[name=promotion_type]:checked', '#promote-form').val()
+      $.ajax({
+        type: 'PUT',
+        url: $(this).attr('action'),
+        dataType: 'json',
+        data: {promotion_type: piece_type, pawn_id: pawnId}
+      }).done(function( data ) {
+        $droppedSpace.html('<span class="piece ui-draggable ui-draggable-handle" data-update-url="/pieces/341" style="position: relative;">'+ data.chess_font_character + '</span>');
+        $('#promote-form').modal('hide');
```
Comment (smlovin2) [inline app/views/games/show.html.erb]: @mscoutermarsh This doesn't end up hiding the modal for some reason. I've tried a couple other similar things but nothing is working.

File: app/views/games/show.html.erb [hunk 2]
```diff
@@ -7,14 +16,38 @@
         $.ajax({
           type: 'PUT',
           url: ui.draggable.data('update-url'),
+          dataType: 'json',
           data: {piece: {row: $(this).parent().index(), column: $(this).index()}}
         }).done(function( data ) {
           $(ui.draggable).detach().css({top: 0, left: 0}).appendTo(drop_space_element);
+          var piece = data.piece
+          var type = data.piece_type
```
Comment (mscoutermarsh) [inline app/views/games/show.html.erb]: missing semi colons on these 2 lines

File: app/views/games/show.html.erb [hunk 3]
```diff
@@ -7,14 +16,38 @@
...
+            $('#pawn-promotion').modal('show');
+          }
         }).fail(function() {
           ui.draggable.animate({top: '0px', left: '0px'}, 'slow');
         });
       }
     });
+
+    $('#promote-form').submit(function(event) {
+      var piece_type = $('input[name=promotion_type]:checked', '#promote-form').val()
```
Comment (mscoutermarsh) [inline app/views/games/show.html.erb]: missing semi colon here

File: app/views/games/show.html.erb [hunk 4]
```diff
@@ -7,14 +16,38 @@
...
+        url: $(this).attr('action'),
+        dataType: 'json',
+        data: {promotion_type: piece_type, pawn_id: pawnId}
+      }).done(function( data ) {
+        $droppedSpace.html('<span class="piece ui-draggable ui-draggable-handle" data-update-url="/pieces/341" style="position: relative;">'+ data.chess_font_character + '</span>');
+        $('#promote-form').modal('hide');
+      }).fail(function() {
+
+      });
+      event.preventDefault();
```
Comment (mscoutermarsh) [inline app/views/games/show.html.erb]: I think you'll want this at the beginning of this method.

File: app/views/games/show.html.erb [hunk 5]
```diff
@@ -7,14 +16,38 @@
...
+
+    $('#promote-form').submit(function(event) {
+      var piece_type = $('input[name=promotion_type]:checked', '#promote-form').val()
+      $.ajax({
+        type: 'PUT',
+        url: $(this).attr('action'),
+        dataType: 'json',
+        data: {promotion_type: piece_type, pawn_id: pawnId}
+      }).done(function( data ) {
+        $droppedSpace.html('<span class="piece ui-draggable ui-draggable-handle" data-update-url="/pieces/341" style="position: relative;">'+ data.chess_font_character + '</span>');
```
Comment (mscoutermarsh) [inline app/views/games/show.html.erb]: should the ID be in `data-update-url`?

File: app/views/games/show.html.erb (hunk 1 above)
Comment (mscoutermarsh) [inline app/views/games/show.html.erb]: Are there any errors on the chrome console when you click it?

**Pull request merged**